astropy/io/votable/tree.py

```diff
@@ -2722,18 +2733,18 @@ def parse(self, iterator, config):
                         self.create_arrays(nrows=0, config=config)
                         return self
 
-        self.create_arrays(nrows=self._nrows, config=config)
-        fields = self.fields
-        names = [x.ID for x in fields]
+        # TODO: make sure all_fields is defined in all cases before this point
+        Field.uniqify_names(all_fields)
+        names = [x.ID for x in all_fields]
         # Deal with a subset of the columns, if requested.
         if not columns:
-            colnumbers = list(range(len(fields)))
+            colnumbers = list(range(len(all_fields)))
```

Comment:
Why is the conversion from `range` to `list` necessary?

Reply:
because `range()` produces a 1-time consumable iterator, but this variable is definitely used more than once.

Reply:
Oh I didn't know that. Was it always true ?

Reply:
As far as I know this was true even for `xrange` in Python 2 (but I can't claim to know Python 2 very well).

I am willing to guess that quite a few people might believe a `range` to be single-use because they know that `range` does lazy evaluation and from that they conclude (mistakenly) that `range` must be a generator.

Reply:
that would make sense, I probably fell in that trap too.